src/bandit-evaluator.ts

```diff
@@ -173,6 +184,43 @@ export class BanditEvaluator {
     subjectKey: string,
     actionWeights: Record<string, number>,
   ): string {
-    return Object.keys(actionWeights)[0]; // TODO: math and stuff
+    // Deterministically "shuffle" the actions
+    // This way as action weights shift, a bunch of users who were on the edge of one action won't all be shifted to the
+    // same new action at the same time.
+    const shuffledActions = Object.entries(actionWeights).sort((a, b) => {
+      const actionAShard = this.sharder.getShard(
+        `${flagKey}-${subjectKey}-${a[0]}`,
+        this.assignmentShards,
+      );
+      const actionBShard = this.sharder.getShard(
+        `${flagKey}-${subjectKey}-${b[0]}`,
+        this.assignmentShards,
+      );
+      let result = actionAShard - actionBShard;
+      if (result === 0) {
+        // In the unlikely case of a tie in randomized assigned shards, break the tie with the action names
+        result = a[0] < b[0] ? -1 : 1;
+      }
+      return result;
+    });
+
+    // Select action from the shuffled actions, based on weight
+    const assignedShard = this.sharder.getShard(`${flagKey}-${subjectKey}`, this.assignmentShards);
+    const assignmentWeightThreshold = assignedShard / this.assignmentShards;
+    let cumulativeWeight = 0;
+    let assignedAction: string | null = null;
+    for (const actionWeight of shuffledActions) {
+      cumulativeWeight += actionWeight[1];
+      if (cumulativeWeight > assignmentWeightThreshold) {
+        assignedAction = actionWeight[0];
+        break;
+      }
+    }
+    if (assignedAction === null) {
```

Comment:
i noticed we tend to do `=== null` and `=== undefined` checks explicitly. any reason to not just do a truthy/falsy check instead, eg.: `if (!assignedAction) {`?

Reply:
In this particular case, there is a (small) possibility that, for whatever reason, a customer uses an action that is the empty string, so I figured I'd be explicit.